### PR TITLE
Override `Task.apply()` to support eager running of Batches

### PR DESCRIPTION
**Pull request by scalen**
Prior to this change, eager running of batches caused errors whereby
the batched task was passed the args and kwargs directly, rather than
wrapping them as a request, when the following was called:
`.apply_async(args, kwargs, *_args, **_kwargs)`

This change updates the minumum viable version of celery, and overrides
the new `Task.apply()` method to pass the request wrapped in a
`SimpleRequest` and a `list` to the underlying function when in eager mode.

### CODE REVIEW

File: requirements/default.txt [hunk 1]
```diff
@@ -1 +1 @@
-celery>=4.0,<5.0
+celery>=4.2.1,<5.0
```
Comment (clokep) [inline requirements/default.txt]: Why this change? Do the tests fail without it?

Comment (scalen) [reply requirements/default.txt]: For some reason I had the impression that `apply()` was not present in earlier versions, but (on double checking) I was wrong.  I'll remove this change.

**Comment by clokep** [comment]
On the surface this change looks pretty reasonable. I'll need to compare it with the implementation of a "normal" task and see if it will break at all.

Any way of adding a test for this?

**Comment by scalen** [comment]
I'm not sure, how are the tests run at the moment? Vanilla `pytest`?

**Comment by clokep** [comment]
Pretty much. The current ones are a little weird cause they run a separate worker in a different thread, I think. Which shouldn't be necessary for an eager task though!

**Comment by scalen** [comment]
Yeah, I've been reading about the celery plugin of pytest. It should be
easy enough to decorate a test function with the `task_always_eager`
config. Hopefully I'll have some time to knock it out tomorrow.

File: celery_batches/__init__.py [hunk 2]
```diff
@@ -289,6 +289,21 @@ def task_message_handler(message, body, ack, reject, callbacks, **kw):
 
         return task_message_handler
 
+    def apply(self, args=None, kwargs=None, *_args, **_kwargs):
+        if self._get_app().conf.task_always_eager:
```
Comment (scalen) [inline celery_batches/__init__.py]: When called from `task.apply_async()`, apparently `task.apply()` doesn't get a `self`??? (This is confirmed with a new test in celery-batches, which I haven't added to the PR yet as it is failing).

By this I mean that a call to task `add.defer(1)` will call through to `add.apply(args=(1,), kwargs={}, ...)`, but that will reach this line as `self=(1,), args=(), kwargs={}, ...`???

Comment (clokep) [reply celery_batches/__init__.py]: I'm not sure why that would be happening, sorry. Seems like `apply` is unbound for some reason though...

Looks like the standard version of this does expect self: https://github.com/celery/celery/blob/cf829307991da3815e1f7b105e736d13dbc7a325/celery/app/task.py#L729-L732

Can you push the test so I can see the error stack?

Comment (scalen) [reply celery_batches/__init__.py]: https://github.com/percipient/celery-batches/pull/16/commits/ebf9b20cccc225b73e292bf536e6b4d56a6b8922

Comment (scalen) [reply celery_batches/__init__.py]: I have investigated and found that all methods on the task that are added or overwritten in the Batches class appear to be functions, rather than bound methods...  I have no idea what could be causing this...

Comment (scalen) [reply celery_batches/__init__.py]: I've discovered the problem, will push another commit soon.

Comment (clokep) [reply celery_batches/__init__.py]: Ah good find! Yeah I wish they would just get rid of the old style `Task` class. 👍 I don't believe there was any particular reason one was used over the other, besides that this was based off the old versions that were in the GitHub repo.

**Comment by scalen** [comment]
Hi @clokep, do you know if there is likely to be any move to merge this at some point? (just reflagging in case it has been forgotten, I apologise if there is some process this is moving through outside of GitHub).

**Comment by clokep** [comment]
@scalen Sorry for the delay! Lost track of this over the holidays. I'll try to get to this "soon". 😄 Please feel free to ping me again if you don't hear back soon!

I think the code looks pretty good, but I wanted to test this in our development environment first to make sure the change of the base task doesn't break anything.

**Comment by scalen** [comment]
Makes sense to me 🙂 Take your time, of course, just wanted to make sure it
was still on your radar!

File: celery_batches/__init__.py [hunk 3]
```diff
@@ -289,6 +288,21 @@ def task_message_handler(message, body, ack, reject, callbacks, **kw):
...
```
Comment (clokep) [inline celery_batches/__init__.py]: Quick question about this, looking at the corresponding code in Celery's `Task` class (see https://github.com/celery/celery/blob/d0563058f8f47f347ac1b56c44f833f569764482/celery/app/task.py#L729-L780), it seems that `apply` is meant to always return an `EagerResult`. This if-statement does, however, exist as part of `apply_async` method: https://github.com/celery/celery/blob/d0563058f8f47f347ac1b56c44f833f569764482/celery/app/task.py#L545-L562

Is this if-statement necessary here? It seems like we should just return `SimpleRequest`? I might be missing something in the way the strategy calls back into `apply()` though...

Comment (scalen) [reply celery_batches/__init__.py]: This `if`-statement is potentially tautologous, yes... I haven't been able to satisfy myself that it isn't called elsewhere, without the `task_always_eager` guard.

On the other hand, having been away and giving it a fresh look now: Even if it were called elsewhere, surely we still want to ensure that the original args/kwargs are wrapped as a `SimpleRequest` before calling the actual task... So yeah, lets get rid of the `if`!

Comment (clokep) [reply celery_batches/__init__.py]: Cool! 👍 I did some testing with our application and this seems to work fine! Sorry for the delay.

Comment (scalen) [reply celery_batches/__init__.py]: Got rid of the `if` 😄

**Comment by clokep** [comment]
Something shook loose while reviewing this again and I realized I had worked on this previously! See 80609f8917e95f0e88c4d0de4f313148b27bb96c for what I had come up with. It is very similar, but I think the key you unlocked was inheriting from the *correct* Task class!

I'm testing this out, but I think it looks good overall! 👍

**Comment by clokep** [comment]
OK! This looks good. I have a couple of commits to add, but GitHub is complaining about me pushing to your branch, so I'll add them on top afterward.

**Comment by clokep** [comment]
See #18 for the changes I made.
